cli/src/cli_util.rs

```diff
@@ -367,7 +373,40 @@ impl CommandHelper {
     #[instrument(skip(self, ui))]
     pub fn workspace_helper(&self, ui: &Ui) -> Result<WorkspaceCommandHelper, CommandError> {
         let mut workspace_command = self.workspace_helper_no_snapshot(ui)?;
-        workspace_command.maybe_snapshot(ui)?;
+
+        let workspace_command = match workspace_command.maybe_snapshot_impl(ui) {
+            Ok(()) => workspace_command,
+            Err(SnapshotWorkingCopyError::Command(err)) => return Err(err),
+            Err(SnapshotWorkingCopyError::StaleWorkingCopy(err)) => {
+                let auto_update_stale = self
+                    .settings()
+                    .config()
+                    .get_bool("snapshot.auto-update-stale")?;
+                if !auto_update_stale {
+                    return Err(err);
+                }
+
+                // We detected the working copy was stale and the client is configured to
+                // auto-update-stale, so let's do that now. We need to do it up here, not at a
+                // lower level (e.g. inside snapshot_working_copy()) to avoid recursive locking
+                // of the working copy.
+                let (workspace_command, status) = self.recover_stale_working_copy(ui)?;
+                match status {
+                    StaleWorkingCopyStatus::NotStale
+                    | StaleWorkingCopyStatus::RecoveredMissingOp => {}
+                    StaleWorkingCopyStatus::RecoveredSnapshot(id) => {
+                        writeln!(
+                            ui.warning_default(),
+                            "Automatically updated to fresh commit {}",
```

Comment:
nit: It's odd that warning is printed only for (less scary) `RecoveredSnapshot(_)` case, though I understand that's probably because we've printed lengthy status message for the `RecoveredMissingOp` case.

Maybe we can move all status/warning messages to `recover_stale_working_copy()` function instead of returning `StaleWorkingCopyStatus`? I thought `workspace_helper()` would call `recover_stale_working_copy()` unconditionally, but actually we check the snapshot error first. So I think it's safe to print `"Nothing to do .."` there.

Reply:
Looks better, thanks. Maybe we can insert a warning saying "recovering from stale working copy automatically..." or something, but it might be too verbose.

Reply:
I'll leave this as-is for now, we can iterate on it after there's some usage.